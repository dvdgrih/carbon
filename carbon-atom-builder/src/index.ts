export * from './schema/CarbonAtomSchema';
export * from './schema/CarbonXmlSchemaParser';
export * from './models/CarbonSchemaModel';
export * from './models/AbstractCarbonAtomModel';
export * from './utils';
