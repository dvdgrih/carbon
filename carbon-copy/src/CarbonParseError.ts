export class CarbonParseError extends Error {
  constructor(error: string) {
    super(error);
  }
}
