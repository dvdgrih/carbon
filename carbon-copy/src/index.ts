export * from './CarbonParseError';
export * from './serializers';

// json
export * from './json/CarbonJsonSerializer';
export * from './json/serialize-definitions';

// xml
export * from './xml/CarbonXMLParseError';
export * from './xml/CarbonXMLSerializer';
export * from './xml/SimpleCarbonAtomSerializer';

export * from './helpers';
