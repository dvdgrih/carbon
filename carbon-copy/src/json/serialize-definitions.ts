export interface Point {
  x: number;
  y: number;
}

export interface MoleculeSerialized {
  atoms: ElementSerialized[];
  annotations: ElementSerialized[];
  links: AtomLinkSerialized[];
}

export interface ElementSerialized extends Point {
  id: string;
  type: string;
}

export interface AtomLinkSerialized {
  target: AtomPortSerialized;
  source: AtomPortSerialized;
  points: Point[];
}

export interface AtomPortSerialized {
  atom_id: string;
  port_key: string;
}
