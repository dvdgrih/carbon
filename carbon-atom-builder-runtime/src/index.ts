import { CarbonSchemaModel } from '@journeyapps-labs/carbon-atom-builder';
import { AtomUIBank } from '@journeyapps-labs/carbon-ui';
import { CustomCarbonAtomModel } from './CustomCarbonAtomModel';

export * from './CustomCarbonAtomModel';
export * from './CustomCarbonAtomFactory';
export * from './CustomCarbonRuntimeAtomSchema';
export * from './CustomCarbonRuntimeAtom';

export const generateAtomUIBank = (model: CarbonSchemaModel) => {
  const bank = new AtomUIBank();
  for (let m of model.atoms) {
    if (m instanceof CustomCarbonAtomModel) {
      bank.registerUI(m.generateAtomUI());
    }
  }
  return bank;
};
