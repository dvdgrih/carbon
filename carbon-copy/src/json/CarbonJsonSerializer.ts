import { AtomFactory, Molecule } from '@journeyapps-labs/carbon-core';
import { Atom, BaseElement } from '@journeyapps-labs/carbon-core';
import { ElementSerialized, MoleculeSerialized } from './serialize-definitions';
import { AttributePort, FlowPort } from '@journeyapps-labs/carbon-core';

export interface JsonElementSerializer<T extends BaseElement = BaseElement> {
  type: string;
  serializeJsonElement(element: T): object;
  deSerializeJsonElement(element: T, data: ReturnType<this['serializeJsonElement']>);
}

export type AtomFactorySerializer = AtomFactory & JsonElementSerializer;

export interface CarbonJsonSerializerOptions {
  factories: AtomFactory[];
  serializers?: JsonElementSerializer[];
}

export class CarbonJsonSerializer {
  factories: Map<string, AtomFactory>;
  serializers: Map<string, JsonElementSerializer>;

  constructor(protected options: CarbonJsonSerializerOptions) {
    this.factories = new Map<string, AtomFactory>();
    this.serializers = new Map<string, JsonElementSerializer>();
    options.factories.forEach((f) => {
      this.factories.set(f.type, f);
    });
    options.serializers?.forEach((f) => {
      this.serializers.set(f.type, f);
    });
  }

  deSerializeMolecule(data: MoleculeSerialized) {
    const molecule = new Molecule();
    const atomMap = new Map<string, Atom>();
    data.atoms.forEach((a) => {
      const atom = this.deSerializeElement(a);
      atomMap.set(atom.id, atom);
      molecule.addAtom(atom);
    });
    data.links.forEach((link) => {
      const port = atomMap.get(link.target.atom_id).getInPort(link.target.port_key);
      if (!port) {
        throw new Error(`broken link [${link.target.atom_id}][${link.target.port_key}]`);
      }
      if (port instanceof AttributePort || port instanceof FlowPort) {
        port.link(atomMap.get(link.source.atom_id).getOutPort(link.source.port_key));
        port.points = link.points || [];
      } else {
        throw new Error(`Unknown port type [${port.atom.id}][${port.key}]`);
      }
    });
    return molecule;
  }

  serializeMolecule(molecule: Molecule): MoleculeSerialized {
    return {
      atoms: Array.from(molecule.atoms.values()).map((a) => {
        return this.serializeElement(a);
      }),
      annotations: Array.from(molecule.annotations.values()).map((a) => {
        return this.serializeElement(a);
      }),
      links: Array.from(molecule.atoms.values()).flatMap((a) => {
        return this.serializeLinks(a);
      })
    };
  }

  serializeLinks(atom: Atom) {
    return atom.getOutPorts().flatMap((sourcePort) => {
      if (sourcePort instanceof AttributePort) {
        return Array.from(sourcePort.childrenPorts.values()).map((port) => {
          return {
            source: {
              atom_id: atom.id,
              port_key: sourcePort.key
            },
            target: {
              atom_id: port.atom.id,
              port_key: port.key
            },
            points: port.points
          };
        });
      } else if (sourcePort instanceof FlowPort) {
        if (!sourcePort.linked) {
          throw new Error(`atom ${atom.id} with out port: ${sourcePort.key} is not linked to anything`);
        }
        return [
          {
            source: {
              atom_id: atom.id,
              port_key: sourcePort.key
            },
            target: {
              port_key: sourcePort.linked.key,
              atom_id: sourcePort.linked.atom.id
            },
            points: sourcePort.linked.points
          }
        ];
      }
    });
  }

  deSerializeElement(data: ElementSerialized) {
    const atom = this.factories.get(data.type).generateElement();
    atom.x = data.x;
    atom.y = data.y;
    atom.id = data.id;
    if (this.serializers.has(data.type)) {
      this.serializers.get(data.type).deSerializeJsonElement(atom, data);
    }
    return atom;
  }

  serializeElement(atom: BaseElement): ElementSerialized {
    let serialized = {
      id: atom.id,
      x: atom.x,
      y: atom.y,
      type: atom.type
    };

    if (this.serializers.has(atom.type)) {
      serialized = {
        ...serialized,
        ...this.serializers.get(atom.type).serializeJsonElement(atom)
      };
    }
    return serialized;
  }
}
