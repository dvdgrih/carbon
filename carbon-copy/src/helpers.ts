import { BaseElement } from '@journeyapps-labs/carbon-core';
import { JsonElementSerializer } from './json/CarbonJsonSerializer';
import { SimpleCarbonAtomSerializer } from './xml/SimpleCarbonAtomSerializer';

export const generateSerializers = <T extends object = {}, A extends BaseElement = BaseElement>(
  type: string,
  serialize: (atom: A) => T,
  deserialize: (atom: A, data: T) => any
) => {
  const json = {
    type: type,
    serializeJsonElement(atom: A) {
      return serialize(atom) as T;
    },
    deSerializeJsonElement(atom: A, data: T): any {
      deserialize(atom, data);
    }
  } as JsonElementSerializer<A>;
  return {
    json: json,
    xml: new SimpleCarbonAtomSerializer(json)
  };
};
