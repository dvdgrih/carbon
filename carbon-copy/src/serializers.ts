import {
  JoinAtom,
  JoinAtomFactory,
  JoinAtomType,
  OutputAtomFactory,
  ParamAtom,
  ParamAtomFactory,
  SetVariableAtomFactory,
  GetVariableAtom,
  GetVariableAtomFactory,
  TextAnnotationElementFactory,
  TextAnnotationElement
} from '@journeyapps-labs/carbon-core';
import { generateSerializers } from './helpers';
import * as _ from 'lodash';
import { XmlAtomSerializer, XmlElementSerializer } from './xml/CarbonXMLSerializer';
import { JsonElementSerializer } from './json/CarbonJsonSerializer';
import { XMLElement, XMLDocument } from '@journeyapps/domparser';

export const generateParamAtomSerializers = (type: string) => {
  return generateSerializers<{ param: string; name: string }, ParamAtom>(
    type,
    (atom) => {
      return {
        param: atom.paramType,
        name: atom.paramName
      };
    },
    (atom, data) => {
      atom.paramType = data.param;
      atom.paramName = data.name;
    }
  );
};

export const generateJoinAtomSerializers = () => {
  return generateSerializers<{ join: JoinAtomType }, JoinAtom>(
    JoinAtomFactory.TYPE,
    (atom) => {
      return {
        join: atom.joinType
      };
    },
    (atom, data) => {
      atom.joinType = data.join;
    }
  );
};

export const generateGetVariableAtomSerializers = () => {
  return generateSerializers<{ variable: string }, GetVariableAtom>(
    GetVariableAtomFactory.TYPE,
    (atom) => {
      return {
        variable: atom.variable
      };
    },
    (atom, data) => {
      atom.variable = data.variable;
    }
  );
};

export const generateTextAnnotationAtomSerializers = (): {
  xml: XmlElementSerializer<TextAnnotationElement>;
  json: JsonElementSerializer<TextAnnotationElement>;
} => {
  return {
    json: {
      type: TextAnnotationElementFactory.TYPE,
      serializeJsonElement(element): object {
        return {
          text: element.text
        };
      },
      deSerializeJsonElement(element, data: any) {
        element.text = data.text || '';
      }
    },
    xml: {
      type: TextAnnotationElementFactory.TYPE,
      serializeXmlElement(node: XMLElement, element, document: XMLDocument): XMLElement {
        node.textContent = element.text;
        return node;
      },
      deSerializeXmlElement(node: XMLElement, element) {
        element.text = node.textContent || '';
      }
    }
  };
};

export const serializerBank: () => { xml: XmlElementSerializer; json: JsonElementSerializer }[] = _.memoize(() => [
  generateParamAtomSerializers(ParamAtomFactory.TYPE),
  generateParamAtomSerializers(OutputAtomFactory.TYPE),
  generateParamAtomSerializers(SetVariableAtomFactory.TYPE),
  generateJoinAtomSerializers(),
  generateGetVariableAtomSerializers(),
  generateTextAnnotationAtomSerializers()
]);

export const generateXMLSerializers: () => XmlAtomSerializer[] = _.memoize(() => {
  return serializerBank().map((s) => s.xml);
});

export const generateJSONSerializers: () => JsonElementSerializer[] = _.memoize(() => {
  return serializerBank().map((s) => s.json);
});
