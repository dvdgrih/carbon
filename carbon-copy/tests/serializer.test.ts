import {
  FlowPort,
  ProgramEndAtom,
  ProgramEndFactory,
  ProgramStartAtom,
  ProgramStartFactory
} from '@journeyapps-labs/carbon-core';
import { CarbonJsonSerializer } from '../src/json/CarbonJsonSerializer';
import { createMolecule } from './fixtures/createMolecule';
import { CarbonXMLSerializer } from '../src';
import { describe, beforeEach, afterEach, it, expect, vi } from 'vitest';

vi.mock('uuid', () => {
  let UUID_ID = 0;

  return {
    v4: () => {
      return `${UUID_ID++}`;
    }
  };
});

describe('Serialization tests', () => {
  const jestConsole = console;

  beforeEach(() => {
    global.console = require('console');
  });

  afterEach(() => {
    global.console = jestConsole;
  });

  it('should serialize a JSON molecule', async () => {
    const serializer = new CarbonJsonSerializer({
      factories: [new ProgramStartFactory(), new ProgramEndFactory()]
    });

    const { molecule } = createMolecule();

    const data = serializer.serializeMolecule(molecule);
    expect(data).toMatchSnapshot();

    const molecule2 = serializer.deSerializeMolecule(data);
    expect(molecule2.getAtomById('0').type).toEqual(ProgramStartFactory.TYPE);
    expect(molecule2.getAtomById('1').type).toEqual(ProgramEndFactory.TYPE);
    expect(molecule2.getAtomById('0').getOutPort<FlowPort>(ProgramStartAtom.PORT_START).linked).toBe(
      molecule2.getAtomById('1').getInPort(ProgramEndAtom.PORT_END)
    );
  });

  it('Should serialize an XML molecule', async () => {
    const serializer = new CarbonXMLSerializer({
      factories: [new ProgramStartFactory(), new ProgramEndFactory()]
    });
    const { molecule, atom_start, atom_end } = createMolecule();
    const data = serializer.serializeMolecule(molecule);
    expect(data).toMatchSnapshot();

    const molecule2 = serializer.deSerializeMolecule(data);
    expect(molecule2.getAtomById(atom_start.id).type).toEqual(ProgramStartFactory.TYPE);
    expect(molecule2.getAtomById(atom_end.id).type).toEqual(ProgramEndFactory.TYPE);
    expect(molecule2.getAtomById(atom_start.id).getOutPort<FlowPort>(ProgramStartAtom.PORT_START).linked).toBe(
      molecule2.getAtomById(atom_end.id).getInPort(ProgramEndAtom.PORT_END)
    );
  });
});
