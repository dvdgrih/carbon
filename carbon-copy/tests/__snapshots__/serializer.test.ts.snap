// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Serialization tests > Should serialize an XML molecule 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<carbon-molecule>
    <atom type="program-start" id="0" x="0" y="0" />
    <atom type="program-end" id="1" x="0" y="0">
        <port name="END" linked-atom="0" linked-port="START" />
    </atom>
</carbon-molecule>
"
`;

exports[`Serialization tests > should serialize a JSON molecule 1`] = `
{
  "annotations": [],
  "atoms": [
    {
      "id": "0",
      "type": "program-start",
      "x": 0,
      "y": 0,
    },
    {
      "id": "1",
      "type": "program-end",
      "x": 0,
      "y": 0,
    },
  ],
  "links": [
    {
      "points": [],
      "source": {
        "atom_id": "0",
        "port_key": "START",
      },
      "target": {
        "atom_id": "1",
        "port_key": "END",
      },
    },
  ],
}
`;

exports[`Serialization tests Should serialize an XML molecule 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<carbon-molecule>
    <atom type=\\"program-start\\" id=\\"0\\" x=\\"0\\" y=\\"0\\" />
    <atom type=\\"program-end\\" id=\\"1\\" x=\\"0\\" y=\\"0\\">
        <port name=\\"END\\" linked-atom=\\"0\\" linked-port=\\"START\\" />
    </atom>
</carbon-molecule>
"
`;

exports[`Serialization tests should serialize a JSON molecule 1`] = `
Object {
  "annotations": Array [],
  "atoms": Array [
    Object {
      "id": "0",
      "type": "program-start",
      "x": 0,
      "y": 0,
    },
    Object {
      "id": "1",
      "type": "program-end",
      "x": 0,
      "y": 0,
    },
  ],
  "links": Array [
    Object {
      "points": Array [],
      "source": Object {
        "atom_id": "0",
        "port_key": "START",
      },
      "target": Object {
        "atom_id": "1",
        "port_key": "END",
      },
    },
  ],
}
`;
